Library/Homebrew/utils/shell.rb

```diff
@@ -0,0 +1,71 @@
+module Utils
+  SHELL_PROFILE_MAP = {
+    :bash => "~/.bash_profile",
+    :csh => "~/.cshrc",
+    :fish => "~/.config/fish/config.fish",
+    :ksh => "~/.kshrc",
+    :sh => "~/.bash_profile",
+    :tcsh => "~/.tcshrc",
+    :zsh => "~/.zshrc",
+  }.freeze
+
+  module Shell
+    # take a path and heuristically convert it
+    # to a shell, return nil if there's no match
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
+      shell_name.to_sym if %w[bash csh fish ksh sh tcsh zsh].include?(shell_name)
+    end
+
+    def self.preferred_shell
+      path_to_shell(ENV.fetch("SHELL", ""))
+    end
+
+    def self.parent_shell
+      path_to_shell(`ps -p #{Process.ppid} -o command=`.chomp)
```

Comment:
I think this should be
```ruby
`ps -p #{Process.ppid} -o ucomm=`.strip
```

Noted with `ucomm=`, there won't be leading - from login shell.